src/core/thread/network_data_local.cpp

```diff
@@ -111,29 +163,91 @@ Error Local::AddOnMeshPrefix(const OnMeshPrefixConfig &aConfig)
     }
 #endif
 
-    error =
-        AddPrefix(aConfig.GetPrefix(), NetworkDataTlv::kTypeBorderRouter, aConfig.mPreference, flags, aConfig.mStable);
-
-exit:
-    return error;
+    return AddPrefix(aConfig.GetPrefix(), NetworkDataTlv::kTypeBorderRouter, aConfig.mPreference, flags,
+                     aConfig.mStable);
 }
 
 Error Local::RemoveOnMeshPrefix(const Ip6::Prefix &aPrefix)
 {
     return RemovePrefix(aPrefix, NetworkDataTlv::kTypeBorderRouter);
 }
 
+Error Local::ValidateHasRoutePrefix(const ExternalRouteConfig &aConfig)
+{
+    Error error = kErrorNone;
+
+    SuccessOrExit(error = ValidatePrefix(aConfig.GetPrefix()));
+    SuccessOrExit(error = ValidatePreference(aConfig.mPreference));
+
+    // Check if the prefix overlaps mesh-local prefix.
+    VerifyOrExit(!aConfig.GetPrefix().ContainsPrefix(Get<Mle::MleRouter>().GetMeshLocalPrefix()),
+                 error = kErrorInvalidArgs);
+
+    // Verify network data contains enough space.
+    VerifyOrExit(CheckBufferSizeForPrefix(aConfig.GetPrefix(), NetworkDataTlv::kTypeHasRoute) == kErrorNone,
+                 error = kErrorInvalidArgs);
```

Comment:
When BBR is not enabled, the domain prefix is not actually added to the `Network Data`. That is why we called `AddOnMeshPrefix` and `RemoveOnMeshPrefix` for `Domain Prefix` just to check if the prefix is valid. 

So, it seems unnecessary to check for buffer size for because:
1. If the BBR is not enabled, setting domain prefix consumes no space in `Network Data`
2. If the BBR is enabled, `AppendTlv` already checked the `Network Data` space

So, I think we should remove calling `CheckBufferSizeForPrefix` in both `ValidateOnMeshPrefix` and `ValidateOnMeshPrefix`.

I am sorry for the confusion here. It has been quite a while since this code was written and the considerations back then is  unclear by now.

Thoughts? @abtink @XPAWAN

Reply:
`CheckBufferSizeForPrefix()` has been removed.